include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_histogram.h

```diff
@@ -0,0 +1,547 @@
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _ONEDPL_PARALLEL_BACKEND_SYCL_HISTOGRAM_H
+#define _ONEDPL_PARALLEL_BACKEND_SYCL_HISTOGRAM_H
+
+#include <algorithm>
+#include <cstdint>
+#include <type_traits>
+
+#include "sycl_defs.h"
+#include "parallel_backend_sycl_utils.h"
+#include "execution_sycl_defs.h"
+#include "unseq_backend_sycl.h"
+#include "utils_ranges_sycl.h"
+#include "oneapi/dpl/internal/histogram_binhash_utils.h"
+#include "oneapi/dpl/internal/async_impl/async_impl_hetero.h"
+
+namespace oneapi
+{
+namespace dpl
+{
+namespace __par_backend_hetero
+{
+
+//no boost baseline
+template <typename _BinHash>
+struct __binhash_SLM_wrapper
+{
+    _BinHash __bin_hash;
+    __binhash_SLM_wrapper(_BinHash __bin_hash) : __bin_hash(__bin_hash) {}
+
+    template <typename _T2>
+    inline ::std::uint32_t
+    get_bin(_T2&& __value) const
+    {
+        return __bin_hash.get_bin(::std::forward<_T2>(__value));
+    }
+
+    template <typename _T2>
+    inline bool
+    is_valid(_T2&& __value) const
+    {
+        return __bin_hash.is_valid(::std::forward<_T2>(__value));
+    }
+
+    inline ::std::size_t
+    get_required_SLM_memory() const
+    {
+        return 0;
+    }
+
+    inline void
+    init_SLM_memory(void* __boost_mem, const sycl::nd_item<1>& __self_item) const
+    {
+    }
+
+    template <typename _T2>
+    inline ::std::uint32_t
+    get_bin(_T2&& __value, void* __boost_mem) const
+    {
+        return get_bin(::std::forward<_T2>(__value));
+    }
+
+    template <typename _T2>
+    inline bool
+    is_valid(_T2&& __value, void* __boost_mem) const
+    {
+        return is_valid(::std::forward<_T2>(__value));
+    }
+};
+
+template <typename _Range>
+struct __binhash_SLM_wrapper<oneapi::dpl::__internal::__custom_range_binhash<_Range>>
+{
+    using _BinHashType = typename oneapi::dpl::__internal::__custom_range_binhash<_Range>;
+    _BinHashType __bin_hash;
+
+    using __boundary_type = typename _BinHashType::__boundary_type;
+
+    __binhash_SLM_wrapper(_BinHashType __bin_hash) : __bin_hash(__bin_hash) {}
+
+    template <typename _T2>
+    inline ::std::uint32_t
+    get_bin(_T2&& __value) const
+    {
+        return __bin_hash.get_bin(::std::forward<_T2>(__value));
+    }
+
+    template <typename _T2>
+    inline bool
+    is_valid(_T2&& __value) const
+    {
+        return __bin_hash.is_valid(::std::forward<_T2>(__value));
+    }
+
+    inline ::std::size_t
+    get_required_SLM_memory()
+    {
+        return sizeof(__boundary_type) * __bin_hash.__boundaries.size();
+    }
+
+    inline void
+    init_SLM_memory(void* __boost_mem, const sycl::nd_item<1>& __self_item) const
+    {
+        ::std::uint32_t __gSize = __self_item.get_local_range()[0];
+        ::std::uint32_t __self_lidx = __self_item.get_local_id(0);
+        ::std::uint8_t __factor = oneapi::dpl::__internal::__dpl_ceiling_div(__bin_hash.__boundaries.size(), __gSize);
```

Comment:
Is there any guarantee that this value will fit within 8 bits? I can imagine a scenario with a small work group size (64) and a decent amount of bins (32,768) which would cause this value to not be represented in a `uint8_t`.